aten/src/ATen/Storage.h

```diff
@@ -1,53 +1,44 @@
 #pragma once
 
-#include "ATen/Scalar.h"
-#include <TH/THStorageFunctions.hpp>
+#include <ATen/StorageImpl.h>
 
 namespace at {
 
-struct Type;
+// OPtions: Use retainable
+// Use strorage ptr == unique ptr <custom destructor calls release on storage>
+//  - custom destructor will just decrease refcount
+// Could rename this to StorageImplImpl and then create StorageImpl that maintains StorageImplImpl and decrease refcount
 
-struct Storage {
-  static const char RESIZABLE = 2;
-
-  Storage() {}
-  Storage(THStorage* storage)
-      : storage(storage) {}
-  Storage(const Storage& other) = delete;
-  virtual ~Storage() {
-    THStorage_free(storage);
-  }
-  void operator=(const Storage&) = delete;
 
-  virtual size_t elementSize() const = 0;
-  size_t size() const {
-    return storage->size;
-  };
-  void* data() {
-    return storage->data_ptr.get();
-  };
-  const void* data() const {
-    return storage->data_ptr.get();
+// StorageImpl/StorageImplImpl thing
+// Inherit from retainable
+// StorageImpl has deleter, ; assignment and move deleted, pretty much like StorageImpl and THStorageImpl before
+// - - don't have to deleted unique ptr - will use intrusive and retainable
+
+
+// Details:
+// Forward the constructor just like before
+// Deconstructor just calls free on StorageImpl
+// Create a StorageImpl free function
+
+struct Storage {
+  Storage() : storage_impl_(nullptr){};
+  Storage(StorageImpl* storage_impl) {
+    storage_impl_ = storage_impl;
+    if (storage_impl_)
+      storage_impl_->retain();
   };
+  ~Storage();
+  Storage(Storage&) = delete;
```

Comment:
Worth a note that this is "temporary", in the sense that there are reasonable implementations of the copy/move constructors/assignment, we just didn't bother putting them in this PR.